File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
from collections import Counter
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        l=0
        count = 0
        check = set()
        for r in range(len(s)):
            while (s[r] in check):
                check.remove(s[l])
                l+=1
            check.add(s[r])
            count=max(r-l+1,count)
        return count

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character, Integer> mp = new HashMap<>();
        int l = 0, maxlen = 0;

        for (int r = 0; r < s.length(); r++) {
            char curr = s.charAt(r);

            if (mp.containsKey(curr) && mp.get(curr) >= l) {
                l = mp.get(curr) + 1;
            }

            mp.put(curr, r);  // ALWAYS update index
            maxlen = Math.max(maxlen, r - l + 1); // ALWAYS update
        }

        return maxlen;
    }
}
